test/utils/test-Variant.cc

```diff
@@ -0,0 +1,895 @@
+
+        THEN("one Bar exists")
+        {
+            REQUIRE(Foo::instances().size() == 0);
+            REQUIRE(Bar::instances().size() == 1);
+            REQUIRE(Baz::instances().size() == 0);
+        }
+
+        THEN("the last Bar is allocated somewhere in the variant")
+        {
+            using address = const std::uint8_t*;
+
+            address v_mem_start = reinterpret_cast<address>(&v);
+            address v_mem_end = v_mem_start + sizeof(v);
+
+            address bar_mem_start = reinterpret_cast<address>(&Bar::last_instance());
+
+            REQUIRE(bar_mem_start >= v_mem_start);
+            REQUIRE(bar_mem_start < v_mem_end);
+        }
+
+        THEN("a Bar was move constructed")
+        {
+            REQUIRE(Bar::record().default_constructor == 0);
+            REQUIRE(Bar::record().copy_constructor == 0);
+            REQUIRE(Bar::record().move_constructor == 1);
+        }
+
+        THEN("no Foo, Baz were constructed")
+        {
+            REQUIRE(Foo::record().default_constructor == 0);
+            REQUIRE(Foo::record().copy_constructor == 0);
+            REQUIRE(Foo::record().move_constructor == 0);
+
+            REQUIRE(Baz::record().default_constructor == 0);
+            REQUIRE(Baz::record().copy_constructor == 0);
+            REQUIRE(Baz::record().move_constructor == 0);
+        }
+
+        THEN("the index of the variant is one")
+        {
+            REQUIRE(v.index() == 1);
+        }
+
+        THEN("get<I=1> on the variant returns the last instance of Bar")
+        {
+            REQUIRE(&get<1>(v) == &Bar::last_instance());
+        }
+
+        THEN("get<I=0,2> on the variant throws bad_variant_access")
+        {
+            REQUIRE_THROWS_AS(get<0>(v), bad_variant_access);
+            REQUIRE_THROWS_AS(get<2>(v), bad_variant_access);
+        }
+
+        THEN("get_if<I=1> on the variant returns the last instance of Bar")
+        {
+            REQUIRE(get_if<1>(v) == &Bar::last_instance());
+        }
+
+        THEN("get_if<I=0,2> on the variant returns nullptr")
+        {
+            REQUIRE(get_if<0>(v) == nullptr);
+            REQUIRE(get_if<2>(v) == nullptr);
+        }
+
+        THEN("holds_alternative<I=1> on the variant is true")
+        {
+            REQUIRE(holds_alternative<1>(v));
+        }
+
+        THEN("holds_alternative<I=0,2> on the variant is false")
+        {
+            REQUIRE(!holds_alternative<0>(v));
+            REQUIRE(!holds_alternative<2>(v));
+        }
+    }
+
+    GIVEN("A variant<Foo, Bar, Baz> initialized from rvalue of variant<Foo, Bar, Baz>(Bar)")
+    {
+        tested bar(in_place_type<Bar>{});
+
+        Foo::reset_record();
+        Bar::reset_record();
+        Baz::reset_record();
+
+        tested v(std::move(bar));
+
+        THEN("one Bar and one Foo exist")
+        {
+            REQUIRE(Foo::instances().size() == 0);
+            REQUIRE(Bar::instances().size() == 2);
+            REQUIRE(Baz::instances().size() == 0);
```

Comment:
The tests and the message is contradicting.

Reply:
:+1 wrongly copy-pasted message